doc/ir-semantics.text

```diff
@@ -121,49 +121,103 @@ declares i =
 requires i =
   | mut x = e
   | const x = e
-  | osr e, _        -> vars(e)
+  | osr (e, v, l, (x = ee)*)
+                    -> vars(e) ∪ vars(ee)*
   | x <- e          -> x :: vars(e)
   | branch x _
   | print x         -> [x]
   | _               -> []
 
-update S i = S \ declares i ∪ requires i
-
-## Well-scoped module
-
-M, S ⊨(scp) l
--------------   ∀ l ∈ M : ∃ S
-     ⊨(scp) P
-
-M, S_l* ⊨(scp) l*
------------------   ∀ l* ∈ pred (M, l) : ∃ S_l* ⊂ update S
-   M, S ⊨(scp) l
-
-
--------------
-M, ∅ ⊨(scp) s
-
-
-## Frame construction
-
-⊨(fra) M    P ⊨(osr) M
-----------------------   ∀ M ∈ P
-       ⊨(fra) P
-
-
-M, S ⊨(fra) l
--------------   ∀ l ∈ M : ∃ S
-     ⊨(fra) P
-
-M, update S ⊨(fra) l*
----------------------   ∀ l* ∈ pred (M, l)
-       M, S ⊨(fra) l
-
-
--------------
-M, ∅ ⊨(fra) s
-
-
-P(v), x̄ ⊨(fra) l
-----------------   ∀ osr(e, v, l, (x = ee)*) ∈ M
-      P ⊨(osr) M
+scope S ::= x*
+scope assignment A ::= ((v, l) → S)*
+
+## Declarative scoping judgment
+
+This judgment just classifies "possible" scope assignments A for
+a program P. It does not characterize the tightest possible
+assignment¹, as it allows to implicitly drop variables from scope on
+each transition (the ⊇ relations in the (v, l → i) rule). In
+particular, it cannot forbid shadowing, as it is always possible to
+drop all same-named bindings before a declaration.
+
+¹: the tightest possible assignment can be defined impredicatively as
+the pointwise intersection of all valid judgments, which is
+well-defined given that valid scopes are bounded by the set of
+variables defined in the program.
+
+  ∀(v → M) ∈ P,   A, P ⊨d v → M
+  ----------------------------
+            A, P ⊨d
+
+  ∀(l → i) ∈ M,   A, P ⊨d v, l → i
+  -------------------------------
+         A, P ⊨d v → M
+
+  S := A(v,l)
+  requires i ⊆ S
+  ∀l* ∈ pred(P(v),l), A(v,l*) ⊇ S
+  ∀l' ∈ succ(P(v),l), S ∪ declares i ⊇ A(v,l')
+  A, P ⊨d i : S
+  ----------------------------------------
+  A, P ⊨d v, l → i
+
+The judgment (A, P ⊢ i : S) below is the instruction-specific judgment
+that checks the scope restrictions that are specific to each
+instruction, rather than the general control-flow constraints and the
+generic requires/declares handling.
+
+  ----------------
+  A, P ⊨d start : ∅
+
+              A(v, l) = x*
+  ----------------------------------
+  A, P ⊨d osr(e, v, l, (x = ee*)) : S
```

Comment:
The judgment is not a tree in this way. The rule

```
  ∀(l → i) ∈ M:   A, P ⊨d v, l → i
  -------------------------------
         A, P ⊨d v → M
```

does a check on each instruction (and each check looks at all predecessors and successors, but not recursively), and then each of those checks also does a `A, P ⊨ i : A(v,l)` instruction-specific check. So the proof tree for that program would be as follows:

```
P :=
  1: start
  2: mut a = 1
  3: print a

A := 1 -> {}, 2 -> {}, 3 -> {a}

∅ ⊆ A(1)              ∅ ⊆ A(2)                  {a} ⊆ {a}
(pred(1) is ∅)        A(1) ⊇ A(2)               A(2) ⊇ A(3) (wrong!)
A(1) ∪ ∅ ⊇ A(2)       A(2) ∪ {a} ⊇ A(3)         (succ(3) is ∅)
A, P ⊨ start : ∅      A, P ⊨ mut a : ∅          A, P ⊨ print a : {a}
----------------      --------------------      -------------------
A, P ⊨ 1 → start      A, P ⊨ 2 → mut a = 1      A, P ⊨ 3 -> print a
-------------------------------------------------------------------
A, P ⊨
```

There is an issue indeed with the "wrong!" part, I need to add a `declares i` somewhere, will look after lunch.